### PR TITLE
plutus-tx: support typeclasses and add some basic classes

### LINKED ISSUE #1109: Typeclasses
What would it take to support typeclasses?

The problem is not the dictionaries: we can model the dictionary-passing just fine. The problem is the *creation* of the dictionaries. In particular, almost all typeclass instance definitions delegate to other instances defined outside the package, with those in `base` being used particularly often.

We can't see the definitions for the `base` instance dictionaries because (surprise surprise) they're typically not `INLINABLE`. The way we handle this for *functions* in `base` is to replace them with our own versions - hence the Plutus  Tx Prelude - but we can't do that for *instances*.

So we need a way to see the bodies of, realistically, everything in all our dependencies. Which means something like https://github.com/input-output-hk/plutus/issues/982.

### PR DESCRIPTION
**Pull request by michaelpj**
Fixes #1109.

It turns out this works just fine. I'm sorry for putting you all through
the pain of not having typeclasses!

We can't use the classes from `base`, unsurprisingly, due to lack
of unfoldings for the instance definitions.

However, we can define our own replacements, which is a bit of
boilerplate but not too bad.

Delightfully, you seem to generally be able to get away with
newtype-deriving them:
- We *do* handle newtype coercions, so the generated instance
definitions are actually fine.
- They seem to be small enough that GHC always puts in an unfolding for
them even though we can't annotate them.
    - This is a little dicey, but seems to be okay so far.

The class operations themselves annoyingly don't have unfoldings, but we
can recreate the RHSs using GHC's functions. Antother annoyance is #978:
this bites us for e.g. the default method declarations in `Eq`, which
means that `(==)` and `(/=)` are mutually recursive! So you need to put
your `Eq` instances in a separate file from the final use site.

To demonstrate, I added `Eq`, `Ord`, and `Functor` and plumbed them
through. I think the result is a lot better!

We get a ~10% increase in script size and a comparable slowdown. I think
this is somewhat inevitable since we're not taking much advantage of
GHC's specializer, so we are paying a price from passing all the
dictionaries around.

TODO (but probably not in this PR):
- `Semigroup`/`Monoid`
- `Applicative`/`Monad`
    - With `RebindableSyntax` I think we could even do `do` notation!
- `Num`?
    - I'm unsure whether we want this, given that `Num` is kind of a
      dodgy class, but I would like to be able to use `+` and `-`...
    - Possibly we should define our own `Num`-like class.
- More tuple instances
    - I was lazy and only went to 2-tuples.

### CODE REVIEW

**Comment by michaelpj** [comment]
> It turns out this works just fine. I'm sorry for putting you all through the pain of not having typeclasses!

(Actually this is slightly unfair to me - it's only been feasible since the unfoldings work landed.)

File: plutus-tx/src/Language/PlutusTx/Prelude.hs [hunk 1]
```diff
@@ -88,6 +86,208 @@ import           Prelude                    as Prelude hiding (all, any, error,
 -- >>> :set -XNoImplicitPrelude
 -- >>> import Language.PlutusTx.Prelude
 
+-- Copied from the GHC definition
+-- | The 'Eq' class defines equality ('==') and inequality ('/=').
+-- All the basic datatypes exported by the "Prelude" are instances of 'Eq',
+-- and 'Eq' may be derived for any datatype whose constituents are also
+-- instances of 'Eq'.
```
Comment (effectfully) [inline plutus-tx/src/Language/PlutusTx/Prelude.hs]: But we do not have deriving, right? I mean, you say we have it, but I don't quite understand how and precisely what.

Comment (michaelpj) [reply plutus-tx/src/Language/PlutusTx/Prelude.hs]: Oops, I just copied the Prelude doc, I should remove the bits that aren't true :sweat_smile:

**Comment by michaelpj** [comment]
I'm going to do a little bit of reorganizing and move the classes and instances into separate files, so don't merge for now.

**Comment by michaelpj** [comment]
Okay, ready for review again.

**Comment by michaelpj** [comment]
One annoyance with this is that we sometimes need to import the Haskell prelude qualified in order to e.g. `fmap` over some types which we don't have instances for in our own prelude.

I'm not really sure what to do there: I kind of *don't* want to add e.g. a `Functor IO` instance in our prelude, given that you can't use it in Plutus Tx...

File: plutus-tx/compiler/Language/PlutusTx/Compiler/Expr.hs [hunk 2]
```diff
@@ -314,7 +316,21 @@ compileExpr e = withContextM 2 (sdToTxt $ "Compiling expr:" GHC.<+> GHC.ppr e) $
...
+        -- Class ops don't have unfoldings in general (although they do if they're for one-method classes, so we
+        -- want to check the unfoldings case first), see the GHC Note [ClassOp/DFun selection] for why. That
+        -- means we have to reconstruct the RHS ourselves, though, which is a pain.
+        GHC.Var n@(GHC.idDetails -> GHC.ClassOpId cls) -> do
+            -- This code (mostly) lifted from MkId.mkDictSelId, which makes unfoldings for those dictionary
+            -- selectors that do have them
+            let sel_names = fmap GHC.getName (GHC.classAllSelIds cls)
+            val_index <- case elemIndex (GHC.getName n) sel_names of
+                Just i  -> pure i
+                Nothing -> throwPlain $ CompilationError "Id not in class method list"
```
Comment (j-mueller) [inline plutus-tx/compiler/Language/PlutusTx/Compiler/Expr.hs]: Can `n` be included in the error?

File: plutus-wallet-api/src/Language/PlutusTx/StateMachine.hs [hunk 3]
```diff
@@ -20,9 +20,8 @@ import           Ledger.Validation         (PendingTx)
 import qualified Ledger.Validation         as V
 
 -- | Specification of a state machine
-data StateMachine s i = StateMachine {
+newtype StateMachine s i = StateMachine {
       smTransition :: s -> i -> s
-    , smStateEq    :: s -> s -> Bool
```
Comment (j-mueller) [inline plutus-wallet-api/src/Language/PlutusTx/StateMachine.hs]: 👍

**Comment by j-mueller** [comment]
> One annoyance with this is that we sometimes need to import the Haskell prelude qualified in order to e.g. fmap over some types which we don't have instances for in our own prelude.

I think that's ok, it won't be something that hits new users right away

File: plutus-tx/compiler/Language/PlutusTx/Compiler/Expr.hs (hunk 2 above)
Comment (michaelpj) [reply plutus-tx/compiler/Language/PlutusTx/Compiler/Expr.hs]: Yep, this is "impossible", but always good to have useful errors when the impossible inevitably happens.

**Comment by michaelpj** [comment]
Verily:
```
michael@clipper:~/projects/iohk/plutus ❯ echo "(program 1.0.0 {(abs a (type) (lam x a x)) (con integer)})" | result/bin/plc evaluate --stdin
(lam x a x)
```

**Pull request merged**